### PR TITLE
Architectural changes

### LINKED ISSUE #1: Redo build scripts
Tasks:

- [ ] remove references to chitech
- [ ] Use find LUA via `find_package`. this is available since cmake 3.18
- [ ] PETSc should be detected using `find_package`. this will need `FindPETSc.cmake`, but we can check PETSc version, etc.
- [ ] all supporting cmake modules are typically stored in `<ROOT>/cmake`, so move stuff there
- [ ] should use `target_link_libraries` and `target_include_directories` instead of `include_directories` and `link_directories` (these are global and really the include and linking stuff is always target specific).
- [ ] binaries should not be produced in the source tree

### PR DESCRIPTION
**Pull request by andrsd**
Changes:
- move `Lua` wrapper into `lua/`
- compile C++ classes as `libopensn`
- compile Lua layer as `libopensnlua`
- redoing cmake scripts
   - using `find_package` to find dependencies with specified versions
   - reducing the number `CMakeLists.txt` there is only 3 now
- adding configure option `OPENSN_WITH_LUA=ON/OFF` (default is ON)
- binary for testing is called `opensn-test`
- binary to run Lua scripts is called `opensn`
- redoing #includes

Closes #1

### CODE REVIEW

**Comment by andrsd** [comment]
Additional comments:
- `LUA_ROOT` is no longer a thing since we are using `find_package`. If libraries are installed in a non-standard location, one uses `CMAKE_PREFIX_PATH=path1:path2:..` to defined the trees where to look for header files and libraries.
- `PETSC_ROOT` still works, but `PETSC_DIR` as well
- as for VTK, no change
- no binaries are produced in a source tree
- config file is now plain `config.h` produced in the binary tree
- no extra `Makefile` is produced in the source tree, there is just the standard one produced by cmake in the binary tree
- `run_tests` should work when executed from the source tree as it did before
  - when executed from the build tree, specifying `--exe test/opensn-test` should work

**Comment by ragusa** [comment]
@andrsd @wdhawkins  the tests are currently failing (quickly). The configure step is actually what is failing. Does this have to do with the class-01 machine and Daryl needs to fix something first ??

File: configure.sh [hunk 1]
```diff
@@ -9,17 +9,17 @@
 #   script depends on the arguments passed:
 #   
 #   1. If called without arguments, CMake scripts for a release (optimized) build
-#      of ChiTech will be generated in the directory chi_build. Additional
+#      of ChiTech will be generated in the directory `build`. Additional
```
Comment (ragusa) [inline configure.sh:12]: ChiTech mention here

**Comment by andrsd** [comment]
@wdhawkins We need to patch the gha workflow for the regression tests. It will need this:
```
- name: Setup environment
  run: |
    echo "CMAKE_PREFIX_PATH=:<PATH WHERE LUA IS>" >> $GITHUB_ENV
```
That should help `cmake` to find Lua

File: configure.sh [hunk 2]
```diff
@@ -9,17 +9,17 @@
...
 #      arguments will be passed to and interpreted by the CMake generator.
 #
 #   2. If called with the single argument from the following list:
 #         Debug, Release, RelWithDebInfo, MinSizeRel,
 #      scripts for the corresponding build of ChiTech will be generated in
```
Comment (ragusa) [inline configure.sh:17]: another mention on this line

**Comment by andrsd** [comment]
Or a similar thing must be done in the module file that load the environment...

File: framework/chi_runtime.cc [hunk 3]
```diff
@@ -157,11 +163,13 @@ Chi::run_time::ParseArguments(int argc, char** argv)
...
+#ifdef OPENSN_WITH_LUA
   if (Chi::run_time::dump_registry_)
   {
     ChiObjectFactory::GetInstance().DumpRegister();
     Chi::console.DumpRegister();
   }
+#endif
 }
 
 // ############################################### Initialize ChiTech
```
Comment (ragusa) [inline framework/chi_runtime.cc:175]: another mention

**Comment by andrsd** [comment]
Comment on the `#include` work. After opensn is installed (which will be a thing in future) via `make install`, it will be sitting in `<INSTALL_PREFIX>/include/opensn` and `<INSTALL_PREFIX>/lib`. Typically as a client app, you only want to specify one include dir which would be `<INSTALL_PREFIX>/include/opensn` with this all the includes will just work, since that 's the common prefix I am assuming in the code.

File: configure.sh (hunk 1 above)
Comment (andrsd) [reply configure.sh:12]: I did not change references to ChiTech in the code, yet. Just in the build script. For code, a lot more work will be required with file renames, etc.

Comment (andrsd) [reply configure.sh:12]: I guess, you see `configure.sh` as part of the build infrastructure, I can add a patch to fix it in this file. For the code changes, I would personally go with a separate PR. IMO PRs should be as small as possible and do just 1 thing.

Comment (andrsd) [reply configure.sh:12]: One more thing, what is final agreement of the capitalization of the code name? IIRC, there were 3 variants:
- OpenSn
- openSn
- opensn

If there is no agreement, yet. No point in changing that in this PR :-)

Comment (ragusa) [reply configure.sh:12]: Yes, I was thinking the .sh can be done now as I've seen you also added opensn so the configure.sh should remove the old name.

File: framework/chi_runtime.cc (hunk 3 above)
Comment (ragusa) [reply framework/chi_runtime.cc:175]: this is a .cc file. it can be done later

File: configure.sh (hunk 1 above)
Comment (ragusa) [reply configure.sh:12]: A non unanimous majority likes: openSn for the code's name so we're going with that.

File: framework/chi_runtime.cc (hunk 3 above)
Comment (andrsd) [reply framework/chi_runtime.cc:175]: We talked about this [here](https://github.com/Open-Sn/openSn/issues/3#issuecomment-1795912565). So I assume a PR where we just fix the comment style first. And another one later with actually changing the names...

File: configure.sh (hunk 1 above)
Comment (andrsd) [reply configure.sh:12]: Ok. I will add a patch to change the code name in `configure.sh`.

**Comment by andrsd** [comment]
I am missing an INCLUDE in the `libopensnlua` target. Will patch...

**Review by ragusa — APPROVED** [review]

**Review by wdhawkins — APPROVED** [review]

**Pull request merged**